### PR TITLE
[posix] fix logging of netlink errors

### PR DESCRIPTION
**Pull request by wgtdkp**
This commit fixes the netlink ACK error logging by extracting error messages from the
`NLMSGERR_ATTR_MSG` attribute. The key point is that the error code in the `nlmsgerr`
header is not necessary `errno` and can't be used with `strerror()`. This is explicitly
specified by the netlink spec ([RFC 3549](https://www.rfc-editor.org/rfc/rfc3549.html#section-2.3.2.2)):

> A non-zero error code message is equivalent to a Negative ACK (NACK).
   In such a situation, the Netlink data that was sent down to the
   kernel is returned appended to the original Netlink message header.
   **An error code printable via the perror() is also set (not in the
   message header, rather in the executing environment state variable)**.

Background:

When `otbr-agent` runs without `CAP_NET_ADMIN`, an error message is printed like below:
> [netif] Failed to process request#1: Unknown error -1

This is because `strerror()` requires a `errno` number of positive value, but existing linux
RTNETLINK implementation seems will return a negative error number. For example:

```c++
        if (kind != RTNL_KIND_GET && !netlink_net_capable(skb, CAP_NET_ADMIN))
                return -EPERM;
```
https://github.com/torvalds/linux/blob/831fe284d8275987596b7d640518dddba5735f61/net/core/rtnetlink.c#L6344-L6345

With the fix in this commit, the error message becomes like below:
> [netif] Failed to process request#1: Operation not permitted

### CODE REVIEW

**Review by abtink — APPROVED** [review]
LGTM. 👍

File: src/posix/platform/netif.cpp [hunk 1]
```diff
@@ -1563,6 +1564,79 @@ static void processNetifInfoEvent(otInstance *aInstance, struct rt_msghdr *rtm)
...
+// ----------------------------------------------
+//
+static void HandleNetlinkResponse(struct nlmsghdr *msg)
+{
+    const struct nlmsgerr *err = reinterpret_cast<const nlmsgerr *>(NLMSG_DATA(msg));
+    size_t                 rtaLength;
+    size_t                 requestPayloadLength = 0;
+    uint32_t               requestSeq           = err->msg.nlmsg_seq;
+
+    if (msg->nlmsg_len < NLMSG_LENGTH(sizeof(struct nlmsgerr)))
```
Comment (bukepo) [inline src/posix/platform/netif.cpp:1593]: If this is truncated, is it safe to use the `requestSeq`?

Comment (bukepo) [inline src/posix/platform/netif.cpp:1593]: If this is truncated, is it safe to use the `requestSeq`?

File: src/posix/platform/netif.cpp [hunk 2]
```diff
@@ -1563,6 +1564,79 @@ static void processNetifInfoEvent(otInstance *aInstance, struct rt_msghdr *rtm)
...
+
+    // The payload of the request is omitted if NLM_F_CAPPED is set
+    if (!(msg->nlmsg_flags & NLM_F_CAPPED))
+    {
+        requestPayloadLength = NLMSG_PAYLOAD(&err->msg, 0);
+    }
+
+    rtaLength = NLMSG_PAYLOAD(msg, sizeof(struct nlmsgerr)) - requestPayloadLength;
+
+    if (err->error == 0)
```
Comment (bukepo) [inline src/posix/platform/netif.cpp:1607]: Is it better to do this check after we verified the message is not truncated?

File: src/posix/platform/netif.cpp [hunk 3]
```diff
@@ -1563,6 +1564,79 @@ static void processNetifInfoEvent(otInstance *aInstance, struct rt_msghdr *rtm)
...
+             rta                = RTA_NEXT(rta, rtaLength))
+        {
+            if (rta->rta_type == NLMSGERR_ATTR_MSG)
+            {
+                errorMsg = reinterpret_cast<const char *>(RTA_DATA(rta));
+                break;
+            }
+            else
+            {
+                otLogInfoPlat("[netif] Ignoring netlink response#%u attribute %d", requestSeq, rta->rta_type);
```
Comment (bukepo) [inline src/posix/platform/netif.cpp:1627]:
```suggestion
                otLogDebgPlat("[netif] Ignoring netlink response#%u attribute %d", requestSeq, rta->rta_type);
```

**Review by bukepo — COMMENTED** [review]
good catch.

File: src/posix/platform/netif.cpp (hunk 1 above)
Comment (wgtdkp) [reply src/posix/platform/netif.cpp:1593]: Thanks for catching this! I moved all initialization to after the truncation check to be consistent (including the pointer cast which is okay before truncation check).

File: src/posix/platform/netif.cpp (hunk 3 above)
Comment (wgtdkp) [reply src/posix/platform/netif.cpp]: done in the new push

File: src/posix/platform/netif.cpp (hunk 2 above)
Comment (wgtdkp) [reply src/posix/platform/netif.cpp:1610]: I think it is already the case. We will `ExitNow()` if we verified that this message is truncated before this line.

Comment (bukepo) [reply src/posix/platform/netif.cpp:1610]: I meant we can verify and exit at line 1600, no need to check NLM_F_CAPPED when `err->error` is `0`.

**Review by bukepo — APPROVED** [review]

File: src/posix/platform/netif.cpp (hunk 2 above)
Comment (wgtdkp) [reply src/posix/platform/netif.cpp]: Got you. Updated in the new push

**Review by jwhui — APPROVED** [review]

**Pull request merged**